src/common/column_matrix.h

```diff
@@ -142,6 +148,118 @@ class ColumnMatrix {
     std::vector<size_t> num_nonzeros;
     num_nonzeros.resize(nfeature);
     std::fill(num_nonzeros.begin(), num_nonzeros.end(), 0);
+
+    if (all_dense) {
+      switch (gmat.index.getBinBound()) {
+        case POWER_OF_TWO_8:
+          SetIndexAllDense(gmat.index.data<uint8_t>(), gmat, nrow);
+          break;
+        case POWER_OF_TWO_16:
+          SetIndexAllDense(gmat.index.data<uint16_t>(), gmat, nrow);
+          break;
+        case POWER_OF_TWO_32:
+          SetIndexAllDense(gmat.index.data<uint32_t>(), gmat, nrow);
+          break;
+      }
+    } else {
+      switch (gmat.index.getBinBound()) {
+        case POWER_OF_TWO_8:
+          SetIndex(gmat.index.data<uint8_t>(), gmat.index.disp(), gmat, nrow, nfeature);
+          break;
+        case POWER_OF_TWO_16:
+          SetIndex(gmat.index.data<uint16_t>(), gmat.index.disp(), gmat, nrow, nfeature);
+          break;
+        case POWER_OF_TWO_32:
+          SetIndex(gmat.index.data<uint32_t>(), gmat.index.disp(), gmat, nrow, nfeature);
+          break;
+      }
+    }
+  }
+
+  /* Fetch an individual column. This code should be used with XGBOOST_TYPE_SWITCH
+     to determine type of bin id's */
+  template <typename T>
+  inline Column<T> GetColumn(unsigned fid) const {
+    Column<T> c(type_[fid],
+                reinterpret_cast<const T*>(&index_[boundary_[fid].index_begin * type_size_]),
+                index_base_[fid], (type_[fid] == ColumnType::kSparseColumn ?
+                &row_ind_[boundary_[fid].row_ind_begin] : nullptr),
+                boundary_[fid].index_end - boundary_[fid].index_begin);
```

Comment:
Please, split this. At least you can use intermediate variables to explain the meaning of the arguments.

Reply:
only type was added :)